forester/src/config.rs

```diff
@@ -1,38 +1,44 @@
 use crate::external_services_config::ExternalServicesConfig;
+use crate::tree_sync::TreeData;
+use chrono::{DateTime, Utc};
 use solana_sdk::pubkey::Pubkey;
 use solana_sdk::signature::Keypair;
+use std::collections::HashMap;
 
 #[derive(Debug)]
 pub struct ForesterConfig {
     pub external_services: ExternalServicesConfig,
-    pub nullifier_queue_pubkey: Pubkey,
-    pub state_merkle_tree_pubkey: Pubkey,
-    pub address_merkle_tree_pubkey: Pubkey,
-    pub address_merkle_tree_queue_pubkey: Pubkey,
     pub registry_pubkey: Pubkey,
     pub payer_keypair: Keypair,
     pub cu_limit: u32,
     pub concurrency_limit: usize,
     pub batch_size: usize,
     pub max_retries: usize,
     pub rpc_pool_size: usize,
+    pub address_tree_data: Vec<TreeData>,
+    pub state_tree_data: Vec<TreeData>,
+    pub last_sync_data: HashMap<Pubkey, TreeSyncMetadata>,
 }
-
 impl Clone for ForesterConfig {
     fn clone(&self) -> Self {
         Self {
             external_services: self.external_services.clone(),
-            nullifier_queue_pubkey: self.nullifier_queue_pubkey,
-            state_merkle_tree_pubkey: self.state_merkle_tree_pubkey,
-            address_merkle_tree_pubkey: self.address_merkle_tree_pubkey,
-            address_merkle_tree_queue_pubkey: self.address_merkle_tree_queue_pubkey,
             registry_pubkey: self.registry_pubkey,
             payer_keypair: Keypair::from_bytes(&self.payer_keypair.to_bytes()).unwrap(),
             cu_limit: self.cu_limit,
             concurrency_limit: self.concurrency_limit,
             batch_size: self.batch_size,
             max_retries: self.max_retries,
             rpc_pool_size: self.rpc_pool_size,
+            state_tree_data: self.state_tree_data.clone(),
+            address_tree_data: self.address_tree_data.clone(),
+            last_sync_data: self.last_sync_data.clone(),
         }
     }
 }
+
+#[derive(Debug, Clone)]
+pub struct TreeSyncMetadata {
+    pub last_synced_at: DateTime<Utc>,
+    pub last_synced_root: [u8; 32],
```

Comment:
what's the conceptual idea behind this?
how often do you sync trees?

Reply:
Removed it for now. 
I sync the trees on every start, but that's obviously not enough, and syncing on every rollover is not enough either. 
I see 2 options:
1. Separate thread that would subscribe to account_compression account changes and do a sync on each change.
2. Just a cron job, every N minutes.

Reply:
Wdyt?

Reply:
The forester epochs and slots will change this a bit. (Foresters are only eligible to forest during their own slots.)
It’s probably enough to sync the tree leading up and during the slot.